### PR TITLE
[breaking] Add prediction fucntion for DMatrix and use inplace predict for dask.

### PR DESCRIPTION
**Pull request by trivialfis**
This PR is about to wrap up https://github.com/dmlc/xgboost/pull/6638, only that `sklearn` is still using `DMatrix` for prediction by default instead of `inplace_predict`.

* Add a new API function for predicting on `DMatrix`.  This function aligns with rest of the `InplacePredictFrom*` functions on semantic of function arguments.
* Purge `ntree_limit` from libxgboost, use iteration instead.
* [dask] Use `inplace_predict` by default for dask sklearn models.
* [dask] Run prediction shape inference on worker instead of client.

The breaking change is in the Python sklearn `apply` function, I made it to be consistent with other prediction functions where `best_iteration` is used by default. Also, close https://github.com/dmlc/xgboost/issues/6553

### CODE REVIEW

**Comment by trivialfis** [comment]
@RAMitchell @ShvetsKS @SmirnovEgorRu This is a big change, could you please help taking a look?

**Comment by trivialfis** [comment]
To make inplace predict the default for sklearn, we need to have some more adapters like coo, csc.

**Comment by trivialfis** [comment]
Something wrong with the docker build.  Please ignore it for review as the last test run was successful.

File: python-package/xgboost/core.py [hunk 1]
```diff
@@ -1553,17 +1602,33 @@ def predict(self,
...
 
+        iteration_range :
+            Specifies which layer of trees are used in prediction.  For example, if a
+            random forest is trained with 100 rounds.  Specifying `iteration_range=(10,
+            20)`, then only the forests built during [10, 20) (half open set) rounds are
+            used in this prediction.
+
+            .. versionadded:: 1.4.0
+
+        strict_shape :
```
Comment (RAMitchell) [inline python-package/xgboost/core.py:1624]: Can you provide a bit more context on why this parameter is needed? Keeping in mind that users can call ndarray.reshape() on the output if they require a strict shape.

File: src/c_api/c_api.cc [hunk 2]
```diff
@@ -604,20 +604,58 @@ XGB_DLL int XGBoosterPredict(BoosterHandle handle,
...
-  learner->Predict(
-      *static_cast<std::shared_ptr<DMatrix>*>(dmat),
-      (option_mask & 1) != 0,
-      &entry.predictions, ntree_limit,
-      static_cast<bool>(training),
-      (option_mask & 2) != 0,
-      (option_mask & 4) != 0,
-      (option_mask & 8) != 0,
-      (option_mask & 16) != 0);
+  ntree_limit = GetLayerTrees(ntree_limit, learner);
```
Comment (RAMitchell) [inline src/c_api/c_api.cc:607]: I'm confused about variable names - this is still ntree_limit, wouldn't this now be called layer_end?

**Review by RAMitchell — COMMENTED** [review]
My understanding of this PR is that the predict c_api is being changed to use a json input, and ntreelimit is being removed in favour of number of iterations.

What happens in the JVM package? I guess it will stay on the old interface until someone updates it.

There are a lot of changes here but most of it is superficial, so it seems generally fine.

File: python-package/xgboost/core.py (hunk 1 above)
Comment (trivialfis) [reply python-package/xgboost/core.py:1624]: Yup.  It's used to clarify the dimensions and make it consistent.  Just to give an example, so on top of mind, could you name the dimension of predict leaf?  Also how many dimensions of softmax should have with and without output margin?

File: src/c_api/c_api.cc (hunk 2 above)
Comment (trivialfis) [reply src/c_api/c_api.cc:607]: That's the old API, I don't want to break it for 1.4.   I will rename the variable.

**Comment by trivialfis** [comment]
> My understanding of this PR is that the predict c_api is being changed to use a json input, and ntreelimit is being removed in favour of number of iterations.

That's right.  A lot of changes are devoted to removing the `ntree_limit` in the codebase.  So the PR looks ugly huge.

> What happens in the JVM package? I guess it will stay on the old interface until someone updates it.

It doesn't use ntree_limit so should be easy to update it.  What's I'm worrying about is the R package, I will make corresponding change, but may need some help from @hcho3 .

File: python-package/xgboost/core.py (hunk 1 above)
Comment (trivialfis) [reply python-package/xgboost/core.py:1624]: At some point we might lower the calculation down into `Learner` class instead of C API.  Right now I just want to remove the burden from users.

**Comment by trivialfis** [comment]
I'm quite excited to remove `ntree_limit`, it's troubling me for a while.

**Review by RAMitchell — APPROVED** [review]

**Comment by RAMitchell** [comment]
@ShvetsKS can you take a look at this also? Thanks

**Comment by ShvetsKS** [comment]
> @ShvetsKS @SmirnovEgorRu would you like to help creating these adapters?

@trivialfis Sure! Sorry for late response. Could you clarify the required time frame for finishing this work?

**Comment by trivialfis** [comment]
> @trivialfis Sure! Sorry for late response. Could you clarify the required time frame for finishing this work?

No problem, I don't have an exact timeline for this, just part of improving the prediction function and memory usage long term roadmap.  I will be working on categorical data and dask for a while so take your time.

At the meanwhile, could you please help taking a look into this PR?  It should set up everything needed for future development.

**Review by wphicks — APPROVED** [review]
This generally looks great! I provided just a couple minor suggestions. The only overall comment that I'd make is that there are quite a lot of unrelated formatting changes in here that make this harder to review. No need to pull them out, but in the future it would probably be better to do a big formatting cleanup as a separate PR.

File: include/xgboost/learner.h [hunk 3]
```diff
@@ -113,8 +113,8 @@ class Learner : public Model, public Configurable, public dmlc::Serializable {
    * \param data input data
    * \param output_margin whether to only predict margin value instead of transformed prediction
    * \param out_preds output vector that stores the prediction
-   * \param ntree_limit limit number of trees used for boosted tree
-   *   predictor, when it equals 0, this means we are using all the trees
+   * \param layer_begin Begining of boosted tree layer used for prediction.
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/learner.h:116]: Good catch.

**Comment by trivialfis** [comment]
> No need to pull them out, but in the future it would probably be better to do a big formatting cleanup as a separate PR.

Thanks for the suggestion!  I tried to go through my PR again, I think some places look like formatting change, but they have actual change like parameter addition.  Feel free to point out where the change is not necessary.  @RAMitchell mentioned formatting issue in my previous PRs a few time. ;-)

File: python-package/xgboost/dask.py [hunk 4]
```diff
@@ -1001,28 +995,22 @@ async def _direct_predict_impl(
...
-        # clear the state to avoid gpu_id, gpu_predictor
-        booster = Booster(model_file=booster.save_raw())
-        test_predt = booster.inplace_predict(test_sample, **kwargs)
-    else:
-        m = DMatrix(test_sample)
-        test_predt = booster.predict(m, **kwargs)
+        kwargs = kwargs.copy()
+        if kwargs["predict_type"] == "margin":
+            kwargs["output_margin"] = True
+        del kwargs["predict_type"]
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply python-package/xgboost/dask.py]: @wphicks Done.

**Comment by trivialfis** [comment]
Close https://github.com/dmlc/xgboost/issues/6621 .

**Comment by ShvetsKS** [comment]
@trivialfis is it expected that with current changes `PredictBatchByBlockOfRowsKernel` function is called even if subsampling is not used (equal to '1')?
It was discovered that during each training iterations `PredictBatchByBlockOfRowsKernel` is called now

**Comment by trivialfis** [comment]
@ShvetsKS I'm not sure how much overhead it has, but I can make a special condition to avoid calling predict if the number of trees for prediction is 0.  Thanks for catching that!   Is there any other concern?

**Comment by trivialfis** [comment]
@ShvetsKS Added a guard.

**Comment by ShvetsKS** [comment]
> @ShvetsKS I'm not sure how much overhead it has, but I can make a special condition to avoid calling predict if the number of trees for prediction is 0. Thanks for catching that! Is there any other concern?

LGTM. No other concerns.

**Comment by trivialfis** [comment]
@RAMitchell @wphicks @ShvetsKS Thanks for the review!

**Comment by wphicks** [comment]
> Thanks for the suggestion! I tried to go through my PR again, I think some places look like formatting change, but they have actual change like parameter addition. Feel free to point out where the change is not necessary. @RAMitchell mentioned formatting issue in my previous PRs a few time. ;-)

Yep, like I said, not a big deal for this PR, and I'm glad it made it in already :tada:! Just for future reference, I was talking about things like [this](https://github.com/dmlc/xgboost/pull/6668/files#diff-b7920fe48897304cecd722fae109a55a22f6fc69b672ab6c81d5ce9d2c6c2556R118-R119). Those single-quotes didn't need to be converted to double-quotes. As a matter of fact, at a quick glance, I'm not sure that that file needed to be touched at all. Are you by chance using some sort of auto-formatter like black or something like that in your dev environment? Anyway, definitely nothing to sweat about :smile:! Just thought I'd mention it since leaving those out will help folks power through review much quicker in the future.

**Comment by trivialfis** [comment]
@wphicks 

> hose single-quotes didn't need to be converted to double-quotes.

Ah ... you are right, that was black.  Will be more careful next time.
